File: packages/python/src/agnes_ai/videos.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any
import time

from ._http import AgnesHTTPClient
from .constants import VIDEO_CREATE_ENDPOINT, VIDEO_MODEL, VIDEO_QUERY_ENDPOINT
from .errors import AgnesAPITimeoutError, AgnesVideoTaskFailedError


class VideosResource:
    def __init__(self, http: AgnesHTTPClient) -> None:
        self._http = http

    def create(
        self,
        *,
        prompt: str,
        model: str = VIDEO_MODEL,
        image: str | list[str] | None = None,
        height: int | None = None,
        width: int | None = None,
        num_frames: int | None = None,
        frame_rate: int | None = None,
        num_inference_steps: int | None = None,
        seed: int | None = None,
        negative_prompt: str | None = None,
        extra_body: Mapping[str, Any] | None = None,
        **extra: Any,
    ) -> dict[str, Any]:
        validate_video_options(num_frames=num_frames, frame_rate=frame_rate)
        body = _drop_none(
            {
                "model": model,
                "prompt": prompt,
                "image": image,
                "height": height,
                "width": width,
                "num_frames": num_frames,
                "frame_rate": frame_rate,
                "num_inference_steps": num_inference_steps,
                "seed": seed,
                "negative_prompt": negative_prompt,
                "extra_body": dict(extra_body or {}) or None,
                **extra,
            }
        )
        return self._http.request("POST", VIDEO_CREATE_ENDPOINT, json_body=body)

    def retrieve(
        self,
        video_id: str,
        *,
        model_name: str | None = None,
    ) -> dict[str, Any]:
        params = _drop_none({"video_id": video_id, "model_name": model_name})
        return self._http.request("GET", VIDEO_QUERY_ENDPOINT, params=params)

    def retrieve_legacy(self, task_id: str) -> dict[str, Any]:
        return self._http.request("GET", f"{VIDEO_CREATE_ENDPOINT}/{task_id}")

    def wait(
        self,
        video_id: str,
        *,
        timeout_seconds: float = 600,
        poll_interval_seconds: float = 5,
    ) -> dict[str, Any]:
        deadline = time.monotonic() + timeout_seconds

        while True:
            data = self.retrieve(video_id)
            status = _get_status(data)

            if status == "completed":
                video_url = extract_video_url(data)
                if video_url is not None:
                    return {**data, "video_url": video_url}
                return data

            if status == "failed":
                raise AgnesVideoTaskFailedError(
                    "Agnes video task failed.",
                    endpoint=VIDEO_QUERY_ENDPOINT,
                )

            if time.monotonic() >= deadline:
                raise AgnesAPITimeoutError(
                    "Timed out waiting for Agnes video task.",
                    endpoint=VIDEO_QUERY_ENDPOINT,
                )

            time.sleep(poll_interval_seconds)


def validate_video_options(
    *,
    num_frames: int | None = None,
    frame_rate: int | None = None,
) -> None:
    if num_frames is not None:
        if num_frames > 441:
            raise ValueError("num_frames must be less than or equal to 441.")
        if (num_frames - 1) % 8 != 0:
            raise ValueError("num_frames must satisfy 8n + 1.")

    if frame_rate is not None and not 1 <= frame_rate <= 60:
        raise ValueError("frame_rate must be between 1 and 60.")


def extract_video_url(data: Mapping[str, Any]) -> str | None:
    video_url = data.get("video_url")
    if isinstance(video_url, str) and video_url:
        return video_url

    remixed_from_video_id = data.get("remixed_from_video_id")
    if isinstance(remixed_from_video_id, str) and remixed_from_video_id:
        return remixed_from_video_id

    return None


def _get_status(data: Mapping[str, Any]) -> str | None:
    status = data.get("status")
    if isinstance(status, str):
        return status.lower()

    nested_data = data.get("data")
    if isinstance(nested_data, Mapping):
        nested_status = nested_data.get("status")
        if isinstance(nested_status, str):
            return nested_status.lower()

    return None


def _drop_none(data: Mapping[str, Any]) -> dict[str, Any]:
    return {key: value for key, value in data.items() if value is not None}

File: packages/javascript/src/videos.ts
import { VIDEO_CREATE_ENDPOINT, VIDEO_MODEL, VIDEO_QUERY_ENDPOINT } from "./constants.js";
import { AgnesVideoTaskFailedError, AgnesAPITimeoutError } from "./errors.js";
import type { JsonObject } from "./http.js";
import { AgnesHTTPClient } from "./http.js";

export interface VideoValidationOptions {
  numFrames?: number;
  frameRate?: number;
}

export interface VideoCreateOptions extends VideoValidationOptions {
  prompt: string;
  model?: string;
  image?: string | string[];
  height?: number;
  width?: number;
  numInferenceSteps?: number;
  seed?: number;
  negativePrompt?: string;
  extraBody?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface VideoWaitOptions {
  timeoutMs?: number;
  pollIntervalMs?: number;
}

export class VideosResource {
  constructor(private readonly http: AgnesHTTPClient) {}

  async create(options: VideoCreateOptions): Promise<JsonObject> {
    validateVideoOptions(options);
    const {
      prompt,
      model = VIDEO_MODEL,
      image,
      height,
      width,
      numFrames,
      frameRate,
      numInferenceSteps,
      seed,
      negativePrompt,
      extraBody,
      ...extra
    } = options;

    return this.http.request("POST", VIDEO_CREATE_ENDPOINT, {
      body: dropUndefined({
        model,
        prompt,
        image,
        height,
        width,
        num_frames: numFrames,
        frame_rate: frameRate,
        num_inference_steps: numInferenceSteps,
        seed,
        negative_prompt: negativePrompt,
        extra_body: extraBody && Object.keys(extraBody).length > 0 ? extraBody : undefined,
        ...extra,
      }),
    });
  }

  retrieve(videoId: string, options: { modelName?: string } = {}): Promise<JsonObject> {
    return this.http.request("GET", VIDEO_QUERY_ENDPOINT, {
      params: {
        video_id: videoId,
        model_name: options.modelName,
      },
    });
  }

  retrieveLegacy(taskId: string): Promise<JsonObject> {
    return this.http.request("GET", `${VIDEO_CREATE_ENDPOINT}/${taskId}`);
  }

  async wait(videoId: string, options: VideoWaitOptions = {}): Promise<JsonObject> {
    const timeoutMs = options.timeoutMs ?? 600_000;
    const pollIntervalMs = options.pollIntervalMs ?? 5_000;
    const deadline = Date.now() + timeoutMs;

    while (true) {
      const data = await this.retrieve(videoId);
      const status = getStatus(data);

      if (status === "completed") {
        const videoUrl = extractVideoUrl(data);
        return videoUrl === undefined ? data : { ...data, video_url: videoUrl };
      }

      if (status === "failed") {
        throw new AgnesVideoTaskFailedError("Agnes video task failed.", {
          endpoint: VIDEO_QUERY_ENDPOINT,
        });
      }

      if (Date.now() >= deadline) {
        throw new AgnesAPITimeoutError("Timed out waiting for Agnes video task.", {
          endpoint: VIDEO_QUERY_ENDPOINT,
        });
      }

      await sleep(pollIntervalMs);
    }
  }
}

export function validateVideoOptions(options: VideoValidationOptions): void {
  const { numFrames, frameRate } = options;

  if (numFrames !== undefined) {
    if (numFrames > 441) {
      throw new RangeError("numFrames must be less than or equal to 441.");
    }
    if ((numFrames - 1) % 8 !== 0) {
      throw new RangeError("numFrames must satisfy 8n + 1.");
    }
  }

  if (frameRate !== undefined && (frameRate < 1 || frameRate > 60)) {
    throw new RangeError("frameRate must be between 1 and 60.");
  }
}

export interface VideoResultLike {
  video_url?: unknown;
  remixed_from_video_id?: unknown;
}

export function extractVideoUrl(data: VideoResultLike): string | undefined {
  if (typeof data.video_url === "string" && data.video_url.length > 0) {
    return data.video_url;
  }

  if (
    typeof data.remixed_from_video_id === "string" &&
    data.remixed_from_video_id.length > 0
  ) {
    return data.remixed_from_video_id;
  }

  return undefined;
}

function getStatus(data: JsonObject): string | undefined {
  if (typeof data.status === "string") {
    return data.status.toLowerCase();
  }

  if (typeof data.data === "object" && data.data !== null && !Array.isArray(data.data)) {
    const nestedData = data.data as Record<string, unknown>;
    if (typeof nestedData.status === "string") {
      return nestedData.status.toLowerCase();
    }
  }

  return undefined;
}

function dropUndefined(data: Record<string, unknown>): JsonObject {
  return Object.fromEntries(Object.entries(data).filter(([, value]) => value !== undefined));
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}
